### DISCUSSION TITLE
Observable - syntax

### DISCUSSION THREAD

**Discussion by redblobgames** [kennyfrc/cami.js#6] [Ideas]
Do you have any thoughts on

    this.count.update(value => value + 1);

vs

    this.count.value = this.count.value + 1

(as used in Vue and Preact)  I think the advantage of the function is that you might want to run it later. So after that statement, this.count might still have the old value. But the advantage of the assignment is that it looks like ordinary variable assignment, and you expect that after that statement, this.count will have the new value. That seems simpler to reason about.

**Comment by kennyfrc** [comment]
Hi @redblobgames! Yes, the value getter / setter syntax was one of my considerations for my API, similar to VanJS, Vue, and Preact.

I ultimately decided on value for the getter and update() for the setter as the syntax got very tricky when it comes to nested reactivity.

For example, stuff like `this.user.value.address.street = "Elm Street"` wouldn't render. A solution was to use structural sharing or just setting `this.user.value = {newObject}` but it's another concept people had to learn. As a result, I believe Vue has two reactive state types: `ref` and `reactive`, one for primitives and another for objects / arrays.

I ultimately decided on just `update()` as it reduces the API that people need to learn and one can freely make ridiculous mutations like below, but while it looks like many fields are set, it does not render each time its set. Instead, it only renders once. 

```
this.user.update(value => {
  if (value.name == 'John') {
    value.name = 'Jane';
    value.age = 31;
    value.address.street = '456 Elm St';
    value.address.city = 'Othertown';
    value.address.country = 'Canada';
    value.address.postalCode = '67890';
    value.address.coordinates.lat = '51.5074';
    value.address.coordinates.long = '0.1278';
  } else {
    value.name = 'John';
    value.age = 30;
    value.address.street = '123 Main St';
    value.address.city = 'Anytown';
    value.address.country = 'USA';
    value.address.postalCode = '12345';
    value.address.coordinates.lat = '40.7128';
    value.address.coordinates.long = '74.0060';
  }
});
 ```

I think the flexibility of allowing people different ways to mutate / change state and have the program "just work" is better than having to learn new conventions. I'm generally optimizing for junior developer experience as I work with those folks a lot.

**Comment by redblobgames** [comment]
> As a result, I believe Vue has two reactive state types: ref and reactive, one for primitives and another for objects / arrays.

BTW, Vue's `ref` works for all values, and`reactive` is a convenient version that doesn't require the extra `.value` everywhere. And I believe Vue will batch up all the mutations and render once in the example you have. So it could be written like this:

```js
const user = reactive(…)

if (user.name == 'John') {
  user.name = 'Jane';
  user.age = 31;
  user.address.street = '456 Elm St';
  user.address.city = 'Othertown';
  user.address.country = 'Canada';
  user.address.postalCode = '67890';
  user.address.coordinates.lat = '51.5074';
  user.address.coordinates.long = '0.1278';
} else {
  user.name = 'John';
  user.age = 30;
  user.address.street = '123 Main St';
  user.address.city = 'Anytown';
  user.address.country = 'USA';
  user.address.postalCode = '12345';
  user.address.coordinates.lat = '40.7128';
  user.address.coordinates.long = '74.0060';
}
```

[Side note: I would prefer that Vue's ref/reactive were merged though, as I'd prefer just one type of observable than two, but … they had to split it because they didn't want to use `this`. In the older class-based API, where they used `this`, they didn't need this split.]

> I'm generally optimizing for junior developer experience as I work with those folks a lot.

I think that's a great design tradeoff. Vue and React seem to be moving away from that, and I've been looking for something simpler.

**Comment by kennyfrc** [comment]
Thanks for that! I'll take a closer look into Vue's implementation and see if there's something we can use.

**Comment by redblobgames** [comment]
It'd be cool if `this` were a vue-like reactive(), so that all modifications are automatically tracked. Then it acts exactly like a regular object, with nothing special (no `.value` like Vue, and no update function like cami). It also works naturally for updating multiple things, and it's uniform whether it's nested or not:

cami
```js
this.location.update(value => 'airplane'); // top level
this.user.update(value => { // nested
   value.status = 'traveling';
});
// two different update() calls
```

vs vue
```js
const location = ref('') // top level uses ref(), can't use reactive()
location.value = 'airplane' // need to use .value with ref()

const user = reactive({status: ''}) // nested uses ref() _or_ reactive()
user.status = 'traveling' // don't use .value with reactive()
```

vs what I would consider the most natural:
```js
// 'this' is a reactive() that has everything, so no need for .value or .update
this.location = 'airplane'; // top level
this.user.status = 'traveling'; // nested
```

However I haven't dug into ES6 Proxy to figure out whether this is possible…

**Comment by kennyfrc** [comment]
@redblobgames it's possible with ES6 Proxies. Here's a spike i had:

Below defines `createObservable`, `track` (dependency tracking), `trigger` (triggers or acts on dependencies), and `effect`.

```javascript
// Reactivity with Proxies
// very similar to Vue's

function isObject(obj) {
    return obj !== null && typeof obj === 'object';
}

function createObservable(obj) {
    if (!isObject(obj)) {
        return obj;
    }

    // Recursively make all properties of the object observable
    for (const key in obj) {
        obj[key] = createObservable(obj[key]);
    }

    return new Proxy(obj, {
        get(target, key, receiver) {
            // Track the property access
            track(target, key);
            // Return the actual value
            return Reflect.get(target, key, receiver);
        },
        set(target, key, value, receiver) {
            // If the new value is an object, make it observable
            if (isObject(value)) {
                value = createObservable(value);
            }
            // Save the old value for comparison
            const oldValue = target[key];
            // Set the new value
            const result = Reflect.set(target, key, value, receiver);
            // If the value has changed, trigger the effects
            if (oldValue !== value) {
                trigger(target, key);
            }
            return result;
        }
    });
}

let currentEffect = null;
const depsMap = new WeakMap();

// Function to track property access
function track(target, key) {
    if (currentEffect) {
        let deps = depsMap.get(target);
        if (!deps) {
            deps = new Map();
            depsMap.set(target, deps);
        }
        let dep = deps.get(key);
        if (!dep) {
            dep = new Set();
            deps.set(key, dep);
        }
        dep.add(currentEffect);
    }
}

// Function to trigger effects
function trigger(target, key) {
    const deps = depsMap.get(target);
    if (deps) {
        const dep = deps.get(key);
        if (dep) {
            dep.forEach(effect => effect());
        }
    }
}

// Function to create an effect
function effect(eff) {
    currentEffect = eff;
    eff();
    currentEffect = null;
}
```

Here's the usage code. The effect is just a simple console.log.

```javascript
const data = createObservable({
    location: 'airplane',
    user: {
        status: 'traveling'
    }
});

effect(() => {
    console.log(`Location: ${data.location}, User Status: ${data.user.status}`);
});

data.location = 'train';
data.user.status = 'working';
```

The constraint with this design is that it only works with objects, if you pass primitives such as strings, numbers, etc, it does not work as javascript primitives are immutable (reactivity can only work with mutable structures like objects). `ref` essentially wraps primitives with an object wrapper that has a `value` key to make it reactive.

I think Vue's new `ref` uses the wrapper object by default, and if it detects the argument is an object, it then uses `reactive()` under the hood ([relevant docs here](https://vuejs.org/guide/essentials/reactivity-fundamentals.html#declaring-reactive-state-1)). Vue also has `shallowRef`, which does not have that `reactive()` fallback for any object that gets passed.

In Vue, this is how normal count & nested object reactivity looks like:

Vue count:
```javascript
// vue
const count = ref(0)

console.log(count) // { value: 0 }
console.log(count.value) // 0

count.value++
console.log(count.value) // 1
```

Vue deep reactivity:
```javascript
// vue
import { ref } from 'vue'

const obj = ref({
  nested: { count: 0 },
  arr: ['foo', 'bar']
})

function mutateDeeply() {
  // these will work as expected.
  obj.value.nested.count++
  obj.value.arr.push('baz')
}
```

And this is how it looks like for Cami (FYI - I plan to release the underlying `ObservableState` object behind Cami in a future release)

Cami counter:
```javascript
this.count = this.observable(0)

this.count.update(value => value + 1);
```

Cami deep reactivity:
```javascript
    this.obj = this.observable({
      nested: { count: 0 },
      arr: ['foo', 'bar']
    });
```

```javascript
    this.obj.update(value => {
      value.nested.count++;
      value.arr.push('baz');
    });
```

The Cami syntax is slightly less natural, but it's is immutable under the hood with immer. It's helpful with undo / redo features and time-travelling through states (haven't exposed the time-travelling features yet, but in the works. Mostly a devtools thing).

FWIW, I think Svelte (before runes) has the most natural API:

Svelte counter:
```javascript
// svelte
let count = 0;

$: console.log(count); // 0

count++; // 1
```

Svelte deep reactivity:
```javascript
// svelte
let obj = {
  nested: { count: 0 },
  arr: ['foo', 'bar']
};

function mutateDeeply() {
  // these will work as expected.
  obj.nested.count++;
  obj.arr.push('baz');
  obj = obj; // This line is necessary to notify Svelte about the change
}

$: console.log(obj); 
```

However, Svelte's ergonomics come at a cost -- they did this at [compile-time](https://svelte.dev/blog/runes#runtime-reactivity). Considering the project's "no build" goal, that highly ergonomic API isn't possible as we do things at runtime.

That said, it's interesting to note that Svelte is now moving away from this and into signal-style API called runes. I believe one motivation for changing this is due to maintainability concerns, as the "it looks like a variable" type of API introduces some readability problems the larger the project is. 

In case you're curious, Vue has an interesting section on [API trade-offs here](https://vuejs.org/guide/extras/reactivity-in-depth.html#api-design-trade-offs).

**Comment by kennyfrc** [comment]
I thought about this a bit more and I'm warming up to the idea of having a more granular state change API to make common tasks more convenient.

For example:

Before:
```javascript
this.count.update(value => value + 1);

this.emailError.update(() => 'Please enter a valid email address.');
```

After:
```javascript
this.count.value++

this.emailError.value = 'Please enter a valid email address.';
```

But beyond that, we can also use object methods / array methods for assigning as well:

Before:
```javascript
this.user.update(value => {
    value.name = this.initialUser.name;
    value.age = this.initialUser.age;
    value.email = this.initialUser.email;
  });

this.tasks.update(tasks => {
  tasks.push(task);
});
```

After:
```javascript
this.user.assign(this.initialUser);

this.tasks.push(task);
```

I actually just pushed a small subset so far to the latest CDN (it doesn't affect existing API). Turns out it wasn't such a big deal to implement as these all used [`update()` under the hood](https://github.com/kennyfrc/cami.js/blob/07fc4d2a1770f16d27e91cde168abd5e4bef7802/src/observable.js#L745). Let me know what you think. The con of the more granular APIs is that it doesn't batch updates like update() -- but it may not matter as much for most applications.

**Comment by redblobgames** [comment]
> The constraint with this design is that it only works with objects, if you pass primitives such as strings, numbers, etc, it does not work as javascript primitives are immutable (reactivity can only work with mutable structures like objects).

I think one possible advantage cami has over vue is that because you provide a ReactiveElement class, you could wrap the *entire* class object inside observable(). Then `this` becomes the object that holds everything. And that simplifies the interface because users don't have to think about primitives vs objects anymore, nor do they have to declare things to be observable(). Everything they put into `this` automatically would be reactive, like magic.

I admit I may be overly optimistic about this. And there's the downside that there's no way to put non-reactive data into the ReactiveElement.

**Comment by kennyfrc** [comment]
That's super interesting -- just had to try it out :) I spiked a very magical API that replaced `customElements.define(...)` with `cami.define(...)` that makes the passed Class auto-inherit from ReactiveElement and wraps it as an observable, but the custom elements protocol didn't like it as it only accepts class definitions. Unfortunately, the wrapper can only be done with instances or proxies (and even if it was feasible, it felt like there's going to be some performance problems with all that wrapping).

```javascript
// the "magical" api. not feasible. also lots of overhead and with the cons you mentioned (i.e. no way to put non-reactive data)
class CounterElement  {
  count = 0

  template() {
    return html`
      <button @click=${() => this.count.value--}>-</button>
      <button @click=${() => this.count.value++}>+</button>
      <div>Count: ${this.count.value}</div>
    `;
  }
}

cami.define('counter-component', CounterElement);
```

However, there's a middle ground that's not so bad I think, which is inspired by [MobX's `makeObservable`](https://mobx.js.org/observable-state.html). We can still define non-reactive data here too. Here, it's called `define` (makeObservable's probably too verbose).

```javascript
class CounterElement extends ReactiveElement {
  count = 0;

  constructor() {
    super();
    this.define({
      observables: ['count']
    });
  }

  template() {
    return html`
      <button @click=${() => this.count.value--}>-</button>
      <button @click=${() => this.count.value++}>+</button>
      <div>Count: ${this.count.value}</div>
    `;
  }
}

customElements.define('counter-component', CounterElement);
```

It's arguably more readable in the long run when you have lots of computed / observable definitions

Before:

```javascript
class CounterElement extends ReactiveElement {
  count = this.observable(0);
  countSquared = this.computed(() => this.count.value * this.count.value);
  countCubed = this.computed(() => this.countSquared.value * this.count.value);
  countSqrt = this.computed(() => Math.sqrt(this.count.value));

// ... other code
```

After: 

```javascript
class CounterElement extends ReactiveElement {
  count = 0;
  countSquared = () => this.count.value * this.count.value;
  countCubed = () => this.countSquared.value * this.count.value;
  countSqrt = () => Math.sqrt(this.count.value);

  // ... other code

  constructor() {
    super();
    this.define({
      observables: ['count'],
      computed: ['countSquared', 'countCubed', 'countQuadrupled', 'countPlusRandom', 'countSqrt']
    })

  // ... rest
```

While similar to MobX, it's still better DX I think as we don't have to define actions, and we don't have to add observer / autorun pass a function component to it like MobX and React. Since we're working with a `ReactiveElement`, there is just a protocol to define a `template` method that returns an html tagged literal and you have everything in one nice class (with the lifecycle callbacks inherited from `HTMLElement`).

**Comment by redblobgames** [comment]
Oh that's really interesting! I hadn't thought about custom elements not going through the proxy.

I knew about MobX but had never looked at it before. Both makeObservable and makeAutoObservable look interesting

**Comment by kennyfrc** [comment]
Dug a little deeper today and it looks like we can do away with `.value` :) Let me know your thoughts on the API below.

We can do a MobX-like convention in the following way (i.e. `define` instead of `makeObservable` and we don't need to define actions on our side. But we'll follow the getter convention for computeds. I think that makes more sense than my earlier function assignment).

```javascript
  class CounterElement extends ReactiveElement {
    count = 0;

    constructor() {
      super();

      this.setup({
        observables: ['count'],
        computed: ['countSquared', 'countCubed', 'countQuadrupled', 'countPlusRandom', 'countSqrt']
      })
    }

    get countSquared() {
      return this.count * this.count;
    }
```

Cami should have other goodies as well like getting values from HTML attributes (just need to define a parse function).

```html
<my-component
    todos='{"data": ["Buy milk", "Buy eggs", "Buy bread"]}'
  ></my-component>
<!-- other stuff -->
<script type="module">
 const { html, ReactiveElement } = cami;

  class MyComponent extends ReactiveElement {
    todos = []

    constructor() {
      super();
      this.setup({
        observables: ['todos'],
        attributes: {
          todos: (v) => JSON.parse(v).data
        }
      });
    }
```

Overall, the future API should feel very natural (see playlist example below). Just "mutate" and it will automatically render. There are a few "gotchas" as it's immutable under the hood, but it only affects nested key updates in large object (for which, the `update` function is useful).

```javascript
  class PlaylistElement extends ReactiveElement {
    playlist = [];

    constructor() {
      super();
      this.setup({
        observables: ['playlist'],
      });
    }

    addSong(song) {
      this.playlist.push(song);
    }

    removeSong(index) {
      this.playlist.splice(index, 1);
    }

    moveSongUp(index) {
      if (index > 0) {
        const songToMoveUp = this.playlist[index];
        const songAbove = this.playlist[index - 1];
        this.playlist.splice(index - 1, 2, songToMoveUp, songAbove);
      }
    }

    moveSongDown(index) {
      if (index < this.playlist.length - 1) {
        const songToMoveDown = this.playlist[index];
        const songBelow = this.playlist[index + 1];
        this.playlist.splice(index, 2, songBelow, songToMoveDown);
      }
    }

    sortSongs() {
      this.playlist.sort();
    }

    reverseSongs() {
      this.playlist.reverse();
    }
```

The eventual trick was to use getter / setter traps (like Vue's `ref`) as the primary interface, then if the underlying data structure is an array or object, you get a Proxy which has a bunch of handlers for property access, functions, etc. The Proxy wraps ObservableState.

```javascript
// snippet
  _handleObjectOrArray(context, key, observable, isAttribute = false) {
    const proxy = this._observableProxy(observable);
    Object.defineProperty(context, key, {
      get: () => proxy,
      set: newValue => {
        observable.update(() => newValue);
        if (isAttribute) {
          this.setAttribute(key, newValue);
        }
      }
    });
  }

  _handleNonObject(context, key, observable, isAttribute = false) {
    Object.defineProperty(context, key, {
      get: () => observable.value,
      set: newValue => {
        observable.update(() => newValue);
        if (isAttribute) {
          this.setAttribute(key, newValue);
        }
      }
    });
  }
```

I haven't released the above yet (will sleep on it). If you don't mind, I'll add you as a co-author in v0.1.0. Your questions / comments have been helpful :)

**Comment by redblobgames** [comment]
Ahh, cool! So the Proxy can't wrap a custom element, but you can insert getter/setter traps for the top level, and use Proxy for the levels underneath? That seems like it should work.

I thought a few nights of sleep would give me more ideas but I don't have a lot to offer right now…

**Comment by kennyfrc** [comment]
>  you can insert getter/setter traps for the top level, and use Proxy for the levels underneath

Yup, you got that right!

**Comment by kennyfrc** [comment]
For those who may be following this thread, we've made another simplification where you don't need to define the constructor anymore, leading to this simpler API:

```
<script src="https://unpkg.com/cami@latest/build/cami.cdn.js"></script>
<script type="module">
  const { html, ReactiveElement } = cami;

  class CounterElement extends ReactiveElement {
    count = 0

    template() {
      return html`
        <button @click=${() => this.count--}>-</button>
        <button @click=${() => this.count++}>+</button>
        <div>Count: ${this.count}</div>
      `;
    }
  }

  customElements.define('counter-component', CounterElement);
</script>
```

More details in this [announcement thread](https://github.com/kennyfrc/cami.js/discussions/9).
